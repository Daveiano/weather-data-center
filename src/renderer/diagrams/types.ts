import {Precision} from "./scaling";
import type {AnnotationMatcher} from "@nivo/annotations";
import {ComputedBarDatum} from "@nivo/bar/dist/types/types";

type DiagramBaseProps = {
  title?: string,
  height: string,
  data: dataItem[],
  precision?: Precision,
  annotations?: AnnotationMatcher<ComputedBarDatum<dataItem>>[]
};

type dataItem = {
  [key: string]: any,
  id: string,
  _id: string,
  time: number,
  timeParsed: string,
  temperature?: number,
  humidity?: number,
  pressure?: number,
  rain?: number,
  solar?: number,
  uvi?: number,
  gust?: number,
  wind?: number,
  wind_direction?: number,
  dew_point?: number,
  felt_temperature?: number,
  temperature_max?: number,
  temperature_min?: number,
  temperature_average?: number
}

export { DiagramBaseProps, dataItem };