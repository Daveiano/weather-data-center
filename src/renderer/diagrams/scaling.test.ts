import {
  getTimeDifferenceInDays,
  scaleAverage,
  scaleMax,
  scaleMin,
  scaleSum,
  scaleMinMaxAvg,
  Precision,
} from "./scaling";
import { dataItem } from "./types";
import scalingInput from "../../../tests/data/scaling-input.json";

const precisionCases: { precision: Precision }[] = [
  { precision: "day" },
  { precision: "week" },
  { precision: "month" },
  { precision: "year" },
];

const sumPrecisionCases: { precision: "week" | "month" | "year" }[] = [
  { precision: "week" },
  { precision: "month" },
  { precision: "year" },
];

it("should get the correct day difference", () => {
  const data: dataItem[] = [
    {
      time: 1614884220,
      timeParsed: "2021-03-04T18:57:00.000Z",
      humidity: 100,
      temperature: 0,
      pressure: 1000,
      _id: "ABC",
      id: "0",
      gust: 0,
      rain: 0,
      uvi: 0,
      dew_point: 0,
      solar: 0,
      wind: 0,
      felt_temperature: 0,
      wind_direction: 0,
    },
    {
      time: 1629041040,
      timeParsed: "2021-08-15T15:24:00.000Z",
      humidity: 100,
      temperature: 0,
      pressure: 1000,
      _id: "DEF",
      id: "1",
      gust: 0,
      rain: 0,
      uvi: 0,
      dew_point: 0,
      solar: 0,
      wind: 0,
      felt_temperature: 0,
      wind_direction: 0,
    },
  ];

  const result = getTimeDifferenceInDays(data);

  expect(result).toBe(163);
});

test.each(precisionCases)("average scale temperature", ({ precision }) => {
  const result = scaleAverage(scalingInput, "temperature", precision);

  expect(result).toMatchSnapshot();
});

test.each(precisionCases)("max scale temperature", ({ precision }) => {
  const result = scaleMax(scalingInput, "temperature", precision);

  expect(result).toMatchSnapshot();
});

test.each(precisionCases)("min scale temperature", ({ precision }) => {
  const result = scaleMin(scalingInput, "temperature", precision);

  expect(result).toMatchSnapshot();
});

test.each(precisionCases)(
  "min/max/average scale temperature",
  ({ precision }) => {
    const result = scaleMinMaxAvg(scalingInput, "temperature", precision);

    expect(result).toMatchSnapshot();
  }
);

test.each(sumPrecisionCases)("sum scale rain", ({ precision }) => {
  const result = scaleSum(scalingInput, "rain", precision);

  expect(result).toMatchSnapshot();
});
