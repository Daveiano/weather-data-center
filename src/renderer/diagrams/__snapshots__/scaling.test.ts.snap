// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should average scale correctly for daily temperature 1`] = `
Array [
  Object {
    "temperature": "1.0",
    "time": 1614884220,
    "timeParsed": "2021-03-04T18:57:00.000Z",
  },
  Object {
    "temperature": "0.0",
    "time": 1614905820,
    "timeParsed": "2021-03-05T00:57:00.000Z",
  },
  Object {
    "temperature": "0.8",
    "time": 1614992220,
    "timeParsed": "2021-03-06T00:57:00.000Z",
  },
  Object {
    "temperature": "1.1",
    "time": 1615078620,
    "timeParsed": "2021-03-07T00:57:00.000Z",
  },
  Object {
    "temperature": "0.5",
    "time": 1615165020,
    "timeParsed": "2021-03-08T00:57:00.000Z",
  },
  Object {
    "temperature": "1.0",
    "time": 1615251420,
    "timeParsed": "2021-03-09T00:57:00.000Z",
  },
  Object {
    "temperature": "2.4",
    "time": 1615337820,
    "timeParsed": "2021-03-10T00:57:00.000Z",
  },
  Object {
    "temperature": "5.0",
    "time": 1615424220,
    "timeParsed": "2021-03-11T00:57:00.000Z",
  },
  Object {
    "temperature": "5.5",
    "time": 1615510620,
    "timeParsed": "2021-03-12T00:57:00.000Z",
  },
  Object {
    "temperature": "6.0",
    "time": 1615595580,
    "timeParsed": "2021-03-13T00:33:00.000Z",
  },
  Object {
    "temperature": "3.8",
    "time": 1615681980,
    "timeParsed": "2021-03-14T00:33:00.000Z",
  },
  Object {
    "temperature": "2.3",
    "time": 1615768380,
    "timeParsed": "2021-03-15T00:33:00.000Z",
  },
]
`;
